### PR TITLE
ENH: b0 reference and skullstrip workflow

### PR DESCRIPTION
**Pull request by josephmje**

### CODE REVIEW

File: dmriprep/interfaces/images.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    out_file = File(exists=True, desc='b0 file')
+
+
+class ExtractB0(SimpleInterface):
+    """
+    Extract all b=0 volumes from a dwi series.
+
+    Example
+    -------
+
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: remove empty line

File: dmriprep/interfaces/images.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+from nipype import logging
+from nipype.interfaces.base import (
+    traits, TraitedSpec, BaseInterfaceInputSpec, SimpleInterface, File
+)
+
+LOGGER = logging.getLogger('nipype.interface')
+
+
+class ExtractB0InputSpec(BaseInterfaceInputSpec):
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion
class _ExtractB0InputSpec(BaseInterfaceInputSpec):
```

File: dmriprep/interfaces/images.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    in_file = File(exists=True, mandatory=True, desc='dwi file')
+    b0_ixs = traits.List(traits.Int, mandatory=True,
+                         desc='Index of b0s')
+
+
+class ExtractB0OutputSpec(TraitedSpec):
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion
class _ExtractB0OutputSpec(TraitedSpec):
```

File: dmriprep/interfaces/images.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    -------
+
+    >>> os.chdir(tmpdir)
+    >>> extract_b0 = ExtractB0()
+    >>> extract_b0.inputs.in_file = str(data_dir / 'dwi.nii.gz')
+    >>> extract_b0.inputs.b0_ixs = [0, 1, 2]
+    >>> res = extract_b0.run()  # doctest: +SKIP
+
+    """
+    input_spec = ExtractB0InputSpec
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion

    input_spec = _ExtractB0InputSpec
```
(one empty line required after the class' docstring)

File: dmriprep/interfaces/images.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    output_spec = ExtractB0OutputSpec
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: output_spec = _ExtractB0OutputSpec

File: dmriprep/interfaces/images.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    """Extract the *b0* volumes from a DWI dataset."""
+    import numpy as np
+    import nibabel as nib
+    from nipype.utils.filemanip import fname_presuffix
+
+    out_file = fname_presuffix(
+        in_file, suffix='_b0', newpath=newpath)
+
+    img = nib.load(in_file)
+    data = img.get_fdata()
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: @arokem can we avoid the internal cast to float64 here? I believe float32 should be enough for the average b0, right?
```suggestion
    data = img.get_fdata(dtype='float32')
```

File: dmriprep/interfaces/images.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+
+class RescaleB0(SimpleInterface):
+    """
+    Rescale the b0 volumes to deal with average signal decay over time
+    and output the median image.
+
+    Example
+    -------
+
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: no newline here

File: dmriprep/interfaces/images.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    hdr = img.header.copy()
+    hdr.set_data_shape(b0.shape)
+    hdr.set_xyzt_units('mm')
+    hdr.set_data_dtype(np.float32)
+    nib.Nifti1Image(b0, img.affine, hdr).to_filename(out_file)
+    return out_file
+
+
+class RescaleB0InputSpec(BaseInterfaceInputSpec):
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion
class _RescaleB0InputSpec(BaseInterfaceInputSpec):
```

(and so on, so forth for the other specs)

File: dmriprep/interfaces/images.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    def _run_interface(self, runtime):
+        self._results['out_file'] = rescale_b0(
+            self.inputs.in_file,
+            self.inputs.mask_file,
+            newpath=runtime.cwd)
+        return runtime
+
+
+def rescale_b0(in_file, mask_file, newpath=None):
+    """
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: remove newline after """

File: dmriprep/interfaces/images.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    Rescale the *b0* volumes from a DWI dataset."""
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: isn't this generic? i.e., this function expects a list of only b=0 volumes

File: dmriprep/interfaces/images.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    import numpy as np
+    import nibabel as nib
+    from nipype.utils.filemanip import fname_presuffix
+
+    out_file = fname_presuffix(
+        in_file, suffix='_median_b0', newpath=newpath)
+
+    img = nib.load(in_file)
+    data = img.get_fdata()
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion
    data = img.get_fdata(dtype='float32')
```

File: dmriprep/interfaces/images.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+    mask_img = nib.load(mask_file)
+    mask_data = mask_img.get_fdata()
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]:
```suggestion
    mask_data = mask_img.get_fdata(dtype='float32')
```

File: dmriprep/interfaces/images.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+    mean_b0_signals = data[mask_data > 0, ...].mean(axis=0)
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: should we go with median here as well?

```suggestion
    mean_b0_signals = np.median(data[mask_data > 0, ...], axis=0)
```

File: dmriprep/interfaces/images.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    Rescale the *b0* volumes from a DWI dataset."""
+    import numpy as np
+    import nibabel as nib
+    from nipype.utils.filemanip import fname_presuffix
+
+    out_file = fname_presuffix(
+        in_file, suffix='_median_b0', newpath=newpath)
+
+    img = nib.load(in_file)
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: we need to check dimensionality here:

```suggestion
    img = nib.load(in_file)
    if  img.dataobj.ndim == 3:
        return in_file
    if img.shape[-1] == 1:
        nb.squeeze_image(img).to_filename(out_file)
        return out_file
```

(please doublecheck the squeeze image function bc I'm not positive it is exactly like that)

File: dmriprep/interfaces/images.py (hunk 10 above)
Comment (josephmje) [reply dmriprep/interfaces/images.py]: Yes, do you mean that the function description should be changed?

File: dmriprep/interfaces/images.py (hunk 14 above)
Comment (josephmje) [reply dmriprep/interfaces/images.py]: good idea! if there's only a single b0, we shouldn't do anything further

File: dmriprep/interfaces/images.py (hunk 13 above)
Comment (oesteban) [reply dmriprep/interfaces/images.py]: hahaha this is what happens when you don't follow the suggestions :P ...

https://354-174237413-gh.circle-artifacts.com/0/tmp/THP002/derivatives/dmriprep/sub-THP0005/log/20200115-044312_27e3f83a-2d31-4dc6-99cb-1a09a8fb857d/crash-20200115-044439-UID1001-rescale_b0-855ed6d2-b9c8-4511-9ff9-529db811dd82.txt

**Comment by josephmje** [comment]
@oesteban thanks for hacking on this! new report! https://458-174237413-gh.circle-artifacts.com/0/tmp/THP002/derivatives/dmriprep/sub-THP0005.html

**Comment by oesteban** [comment]
I tried a fix to the niworkflows problem and then rolled back. I'm going to go for it. Thanks for this precious contribution!

**Pull request merged**